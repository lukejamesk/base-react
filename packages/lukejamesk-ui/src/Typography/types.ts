export type TypographyIntrinsicElements = Required<
  Pick<
    JSX.IntrinsicElements,
    'p' | 'strong' | 'blockquote' | 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' | 'span'
  >
>

export type TypographyVariant = Required<
  | 'paragraph'
  | 'bold'
  | 'blockquote'
  | 'heading1'
  | 'heading2'
  | 'heading3'
  | 'heading4'
  | 'heading5'
  | 'heading6'
  | 'text'
>
