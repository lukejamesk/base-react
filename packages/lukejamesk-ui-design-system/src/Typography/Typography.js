import React, { forwardRef } from 'react'
import PropTypes from 'prop-types'
import clsx from 'clsx'
import { Typography as TypographyCore } from '@lukejamesk/core'

const variantMapping = {
  paragraph: 'ljk-typo-p',
  bold: 'ljk-typo-b',
  blockquote: 'ljk-typo-blockquote',
  heading1: 'ljk-typo-h1',
  heading2: 'ljk-typo-h2',
  heading3: 'ljk-typo-h3',
  text: 'ljk-typo-text',
}

const alignMapping = {
  auto: '',
  left: 'ljk-text-left',
  center: 'ljk-text-center',
  right: 'ljk-text-right',
  justify: 'ljk-text-justify',
}

export const typographyVariants = [
  'paragraph',
  'bold',
  'blockquote',
  'heading1',
  'heading2',
  'heading3',
  'text',
]

export const typographyAlignments = ['auto', 'left', 'center', 'right', 'justify']

const Typography = forwardRef(
  ({ variant = 'text', className, paragraph = false, align = 'auto', ...other }, ref) => {
    const vMapped = paragraph ? 'paragraph' : variant

    return (
      <TypographyCore
        variant={vMapped}
        className={clsx(variantMapping[vMapped], alignMapping[align], className)}
        ref={ref}
        {...other}
      />
    )
  },
)

Typography.displayName = 'Typography'

Typography.propTypes = {
  variant: PropTypes.oneOf(typographyVariants),
  paragraph: PropTypes.bool,
  className: PropTypes.string,
  align: PropTypes.oneOf(typographyAlignments),
}

export default Typography
