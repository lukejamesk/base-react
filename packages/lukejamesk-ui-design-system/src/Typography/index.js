export { default, typographyVariants } from './Typography'
