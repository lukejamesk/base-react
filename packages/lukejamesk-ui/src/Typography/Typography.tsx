import React, { forwardRef } from 'react'
import { TypographyVariant, TypographyIntrinsicElements } from './types'

export const typographyVariants: Record<TypographyVariant, keyof TypographyIntrinsicElements> = {
  paragraph: 'p',
  bold: 'strong',
  blockquote: 'blockquote',
  heading1: 'h1',
  heading2: 'h2',
  heading3: 'h3',
  heading4: 'h4',
  heading5: 'h5',
  heading6: 'h6',
  text: 'span',
}

export type TypographyProps = React.PropsWithChildren<{
  variant?: TypographyVariant
  className?: string | undefined
}>

export type Ref = any

const findVariant = (variant?: TypographyVariant) => {
  return variant ? typographyVariants[variant] : typographyVariants.paragraph
}

const Typography = forwardRef<Ref, TypographyProps>(({ variant, className, ...other }, ref) => {
  const Component: keyof TypographyIntrinsicElements = findVariant(variant)
  return <Component className={className} {...other} ref={ref} />
})

Typography.displayName = 'Typography'

export default Typography
