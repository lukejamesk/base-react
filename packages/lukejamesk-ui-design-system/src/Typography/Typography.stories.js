import React from 'react'
import { text, boolean } from '@storybook/addon-knobs'
import Typography from './Typography'

const exampleText = `Lorem ipsum dolor sit amet, consectetur adipiscing elit.
Maecenas feugiat varius mi, eu sagittis nunc dignissim nec.
Quisque rhoncus hendrerit leo.`

export default {
  title: 'Design System|Typography',
  component: Typography,
}

export const Paragraph = () => <Typography paragraph={boolean}>{text('text', exampleText)}</Typography>
export const Blockquote = () => <Typography variant="blockquote">{text('text', exampleText)}</Typography>
export const Bold = () => <Typography variant="bold">{text('text', exampleText)}</Typography>
export const Heading1 = () => <Typography variant="heading1">{text('text', 'Heading 1')}</Typography>
export const Heading2 = () => <Typography variant="heading2">{text('text', 'Heading 2')}</Typography>
export const Heading3 = () => <Typography variant="heading3">{text('text', 'Heading 3')}</Typography>
