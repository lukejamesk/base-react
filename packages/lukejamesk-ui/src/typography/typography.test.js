import React from 'react'
import { mount } from 'enzyme'
import { forEach } from 'ramda'
import Typography from './typography'

const textTypes = [
  'paragraph',
  'bold',
  'blockquote',
  'heading1',
  'heading2',
  'heading3'
]

describe('<Typography />', () => {
  forEach((type) => {
    it(`should render with: ${type}`, () => {
      const wrapper = mount(
        <Typography type={type}>
          My content for
          {type}
        </Typography>
      )
      expect(wrapper).toMatchSnapshot()
    })
  }, textTypes)

  it('should render a paragraph when paragraph is true', () => {
    const wrapper = mount(
      <Typography paragraph>
        My content for paragraph
      </Typography>
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('with no type', () => {
    const wrapper = mount(
      <Typography>
        My content for paragraph
      </Typography>
    )
    expect(wrapper).toMatchSnapshot()
  })
})
