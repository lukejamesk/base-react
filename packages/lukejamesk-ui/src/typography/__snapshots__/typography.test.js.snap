// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Typography /> should render a paragraph when paragraph is true 1`] = `
<ForwardRef
  paragraph={true}
>
  <p>
    My content for paragraph
  </p>
</ForwardRef>
`;

exports[`<Typography /> should render with: blockquote 1`] = `
<ForwardRef
  type="blockquote"
>
  <blockquote>
    My content for
    blockquote
  </blockquote>
</ForwardRef>
`;

exports[`<Typography /> should render with: bold 1`] = `
<ForwardRef
  type="bold"
>
  <strong>
    My content for
    bold
  </strong>
</ForwardRef>
`;

exports[`<Typography /> should render with: heading1 1`] = `
<ForwardRef
  type="heading1"
>
  <h1>
    My content for
    heading1
  </h1>
</ForwardRef>
`;

exports[`<Typography /> should render with: heading2 1`] = `
<ForwardRef
  type="heading2"
>
  <h2>
    My content for
    heading2
  </h2>
</ForwardRef>
`;

exports[`<Typography /> should render with: heading3 1`] = `
<ForwardRef
  type="heading3"
>
  <h3>
    My content for
    heading3
  </h3>
</ForwardRef>
`;

exports[`<Typography /> should render with: paragraph 1`] = `
<ForwardRef
  type="paragraph"
>
  <p>
    My content for
    paragraph
  </p>
</ForwardRef>
`;

exports[`<Typography /> with no type 1`] = `
<ForwardRef>
  <p>
    My content for paragraph
  </p>
</ForwardRef>
`;
